Add a new page for Chainguard Actions (#3440)
chainguard-dev/edu: content/chainguard/actions/overview.md

---
title: "Chainguard Actions overview"
linktitle: "Overview"
description: "Learn how Chainguard Actions provides hardened drop-in replacements for popular GitHub Actions to protect your CI/CD pipelines from supply chain attacks."
type: "article"
date: 2026-06-18T00:00:00+00:00
lastmod: 2026-06-18T00:00:00+00:00
draft: false
tags: ["Chainguard Actions", "Overview"]
menu:
  docs:
    parent: "actions"
weight: 001
toc: true
---

> Chainguard Actions is a set of hardened drop-in replacements for popular GitHub Actions. Each action preserves the same inputs and outputs as the upstream version, but have been examined and revised to better protect your CI/CD pipelines from supply chain attacks. The only change in your workflow configuration is the name of the action in the `uses:` line.

Reviewer: ```suggestion
Chainguard Actions are a set of hardened drop-in replacements for popular GitHub Actions. Each action preserves the same inputs and outputs as the upstream version, but has been examined and revised to better protect your CI/CD pipelines from supply chain attacks. The only change in your workflow configuration is the name of the action in the `uses:` line.
```

> Coverage spans GitHub first-party (`actions/*`), cloud-provider (`aws-actions/*`, `azure/*`, `google-github-actions/*`), Docker, HashiCorp, security tools (Trivy, Grype, CodeQL, Semgrep), and a growing catalog of community actions.

Reviewer: ```suggestion
Coverage spans GitHub first-party (`actions/*`), cloud-provider (`aws-actions/*`, `azure/*`, `google-github-actions/*`), Docker, HashiCorp, and security tools actions (Trivy, Grype, CodeQL, Semgrep), as well as a growing catalog of community actions.
```

Each hardened action:

- Is built from source and evaluated through a rule-based and AI-powered hardening pipeline

> - Has every internal `uses:` and container image reference pinned to an immutable SHA

Reviewer: ```suggestion
- Has every internal `uses:` and container image reference pinned to an immutable SHA digest
```

maybe?

Reviewer: If it makes sense, i would default to `SHA digest` throughout, but your call

- Ships with a `HARDENING.md` report documenting exactly what was checked and fixed
- Is re-reviewed and re-hardened whenever upstream publishes a new version or Chainguard adds a new rule

> Chainguard Actions protects against common threats including tag hijacking, dependency confusion, `pull_request_target` abuse, and secret exfiltration.

Reviewer: ```suggestion
Chainguard Actions protect against common threats including tag hijacking, dependency confusion, `pull_request_target` abuse, and secret exfiltration.
```

Maybe I'm wrong with these singular vs. plural changes, but grammatically I think we should always treat "Chainguard Actions" as plural

Author: I was on the fence. There's an argument for treating this as a collective noun, but it's borderline. I just picked and tried to be consistent, but doing it as plural also makes sense.

This page provides enough to get you started. See the [Chainguard Actions README](https://github.com/chainguard-actions) in GitHub for deeper technical details and some example migrations.

## Prerequisites

> - An organization in the Chainguard platform that you own or can administer and that is registered to use actions

Reviewer: ```suggestion
- An organization in the Chainguard platform that you own or can administer and that is registered to use Chainguard Actions
```

+ maybe a line directing customers how to get set up w/ Actions? Point them to the product page?

## Preliminary steps

Before using Chainguard Actions, log in to Chainguard and enable the Chainguard Actions entitlement for your organization.

> ### Log in to Chainguard

Reviewer: ```suggestion

```

Authenticate using `chainctl`:

```shell
chainctl auth login
```

Confirm your authentication status:

```shell
chainctl auth status
```

### Enable Chainguard Actions for your organization

> 

Reviewer: ```suggestion
```

Checking auth status seems extraneous to me, and with just one command under 'Log in ...' I don't think we need H3s here

Create the Chainguard Actions entitlement to enable access to the hardened actions hosted at `github.com/chainguard-actions`:

```shell

> chainctl actions entitlements create --parent $ORGANIZATION

Reviewer: ```suggestion
chainctl actions entitlements create
```

We got a suggestion awhile back that we don't need to include --parent in chainctl commands, as most users are only members of a single org

Author: I don't remember why at the moment, but there was a discussion about this in this page's context and the SMEs consensus was it was needed across this page.

```

The output confirms the entitlement:

> ```

Reviewer: ```suggestion
```output
```

Enabled Actions product for org chainguard.edu ($ENTITLEMENT_ID) [entitlement id: $ENTITLEMENT_ID]
```

Confirm your entitlement:

```shell

> chainctl actions entitlements list --parent $ORGANIZATION

Reviewer: ```suggestion
chainctl actions entitlements list
```

```

> ```

Reviewer: ```suggestion
```output
```

```

## Basic usage (quick start)

To use a Chainguard hardened action, edit your workflow's YAML configuration file and change the `uses:` line to match the location in `chainguard-actions`:

```yaml
- uses: chainguard-actions/<action-name>@main
```

Action names often have the upstream organization appended to the action name for clarity, for example, `tj-actions/changed-actions` becomes `tj-actions-changed-actions`. This prevents two different sources of a `changed-actions` action from clashing in the Chainguard Actions repository.

Search the Chainguard Actions repository, find the action you want to use, and then use the name you find there.

> **Note:** This example uses `@main`, a mutable reference, to illustrate the mechanics of switching organizations. For production workflows, pin to an immutable SHA instead. The [Configure your workflows](#configure-your-workflows-to-use-chainguard-actions) section covers the full migration, including how to [find the correct SHA](#find-the-sha-for-a-specific-release).

The rest of this page goes a bit deeper into how to use Chainguard Actions.

## Configure your workflows to use Chainguard Actions

> 1. Inventory the actions you currently use.

Reviewer: I would make all of these steps H3s, it looks a little off as it is

Reviewer: I also think the indentation throughout this section looks a little odd, using H3s instead would mean you could get rid of the indents

   Run this from the root of your repository to get a deduplicated list of every `uses:` line across every workflow:

   ```shell
   grep -rhE "uses:\s*[^@]+@" .github/workflows/ | sort -u
   ```

2. Check the Chainguard Actions catalog for each action.

   Browse [the Chainguard Actions repository](https://github.com/chainguard-actions) or use the GitHub search UI. Match by organization and action name — for example, if you use `tj-actions/changed-files`, search for `org:chainguard-actions tj-actions-changed-files`.

   If the action isn't in the catalog, [open an issue](https://github.com/chainguard-actions/.github/issues/new?template=new-action.yml) to request it.

3. Replace the `uses:` line in each workflow.

>    Change only the organization; keep the action name. Find and pin to the commit SHA and preserve the original tag as a comment so Dependabot, Renovate, and human reviewers can track upgrades:

Reviewer: This is a bit unclear, as you state earlier that the action name will change to reflect the upstream org

   ```yaml
   # Before
   - uses: tj-actions/changed-files@v47
   ```

   ```yaml
   # After
   - uses: chainguard-actions/tj-actions-changed-files@<SHA> # v47
   # originally - uses: tj-actions/changed-files@v47
   ```

>   **Find the SHA for a specific release**

Reviewer: ```suggestion
```

I don't think this needs a separate heading

   To find the SHA for a specific release, use the `gh` CLI:

   ```shell
   gh api repos/chainguard-actions/tj-actions-changed-files/commits/v47 --jq '.sha'
   ```

   ```output
   25a1eb5aa40568ec6f8c0e58f2e809ef4270ebfa
   ```

   For the short SHA:

   ```shell
   gh api repos/chainguard-actions/tj-actions-changed-files/commits/v47 --jq '.sha[:7]'
   ```

   ```output
   25a1eb5
   ```

   The resulting `uses:` line with the full SHA:

   ```yaml
   - uses: chainguard-actions/changed-files@25a1eb5aa40568ec6f8c0e58f2e809ef4270ebfa # v47
   ```

4. Update your allowed-actions list.

   If your GitHub organization or repository restricts which actions can run (**Settings > Actions > General > Allow select actions**), add `chainguard-actions/*` to the allowed patterns. Without this, workflows fail with a policy error on first run.

   > **Note:** It's a good idea to remove allowed actions that are no longer being used.

5. Commit, open a PR, and verify that CI passes.

   The action's inputs, outputs, and behavior are almost always identical to the upstream version, so no other workflow changes are needed.

>    > **Note:** Read the `HARDENING.md` file for each Chainguard Action before migrating. In rare cases, the hardening process requires a change to inputs, outputs, or behavior — those changes are documented in this file.

Reviewer: two notes so close to one another looks a little off too, maybe massage one into just being part of the regular body text?

   If something breaks, [file an issue](https://github.com/chainguard-actions/.github/issues/new?template=action-issue.yml) with a reproducer.

## View the actions you are currently using in a repository

Use `chainctl` to scan every workflow and composite action in a repository and list all dependencies transitively:

```shell

> chainctl actions discover $GIT_ORGANIZATION/$REPO

Reviewer: neat!

```

> ```

Reviewer: ```suggestion
```output
```

    scanning $GIT_ORGANIZATION/$REPO for workflows and actions
               ACTION                    | REQUESTED | USED BY
    -------------------------------------|-----------|---------
     actions/checkout                    | v4        | 1
     chainguard-actions/actions-checkout | v6.0.2    | 1

    2 actions, 0 container images

```

## View the actions currently available

While you can search the [Chainguard Actions repository](https://github.com/chainguard-actions) directly in GitHub, you can also use `chainctl` to find an action.

```shell
chainctl actions catalog list --upstream-owner=$OWNER
```

For example, to list all the actions from the `tj-actions` source:

```shell
chainctl actions catalog list --upstream-owner=tj-actions
```

This example returns a list of all actions in the Chainguard Actions repository that originate from the `tj-actions` upstream source.

## Hardened action repository contents

The main branch of each hardened action repository contains:

> - `HARDENING.md` — the authoritative, per-action record of what was checked, what was fixed, and how.

Reviewer: the bullets here all end in periods, while bullets above don't. Not sure what the Google Style guide recommends but I'd go for consistency either way

Reviewer: here's the relevant page, there are a lot of exceptions/rules so use your best judgement: https://developers.google.com/style/lists

- `action.yml` or `action.yaml` — the hardened action definition, preserving upstream inputs and outputs with fixes applied.
- `LICENSE_CHAINGUARD` — the Chainguard license for the hardened variant.
- `source.json` and `published.json` — manifests pointing at the upstream source and the upstream version being tracked. Not yet present in all repos; some older repos don't include them.
- Some actions also include documentation from upstream that you can adapt to use with the Chainguard hardened version.

Then, the version branches in the hardened action repos contain the hardened actions.

## The continuous re-hardening process

Chainguard Actions are continuously re-hardened:

- When upstream publishes a new version, the pipeline re-runs and publishes a new hardened version.
- When the hardening ruleset is updated, affected actions are re-reviewed against the new rules.
- The `HARDENING.md` report is regenerated on every hardening run, with its own policy SHA pinning the exact set of rules that were applied.

> ## Request a new action

Reviewer: these last two sections are so short, and both have the same call to action. I wonder if they could be combined?
